crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -388,87 +570,46 @@ impl<P: RoomDataProvider> TimelineInner<P> {
+            .get(&AnnotationKey::from(annotation))
+            .expect("Reaction state should be set before sending the reaction");
 
-            // Remove the local echo from the related event.
-            if let Some(txn_id) = local_echo_to_remove {
-                if reaction_group.0.remove(&(Some(txn_id.to_owned()), None)).is_none() {
-                    warn!(
-                        "Tried to remove reaction by transaction ID, but didn't \
-                             find matching reaction in the related event's reactions"
-                    );
-                }
+        let follow_up_action = match (result, reaction_state) {
+            (ReactionToggleResult::AddSuccess { event_id, .. }, ReactionState::Redacting(_)) => {
+                // A reaction was added successfully but we've been requested to undo it
+                ReactionAction::RedactRemote(event_id.to_owned())
             }
-            // Add the remote echo to the related event
-            if let Some(event_id) = remote_echo_to_add {
-                let own_user_id = self.room_data_provider.own_user_id().to_owned();
-                reaction_group.0.insert((None, Some(event_id.to_owned())), own_user_id);
-            };
-            if reaction_group.0.is_empty() {
-                reactions.remove(&annotation.key);
+            (ReactionToggleResult::RedactSuccess { .. }, ReactionState::Sending(txn_id)) => {
+                // A reaction was was redacted successfully but we've been requested to undo it
+                ReactionAction::SendRemote(txn_id.to_owned())
             }
-
-            reactions
+            _ => ReactionAction::None,
         };
-        let new_related = related.with_kind(remote_related.with_reactions(new_reactions));
 
-        // Update the reactions stored in the timeline state
-        {
-            // Remove the local echo from reaction_map
-            // (should the local echo already be up-to-date after event handling?)
-            if let Some(txn_id) = local_echo_to_remove {
-                if state.reaction_map.remove(&(Some(txn_id.to_owned()), None)).is_none() {
-                    warn!(
-                        "Tried to remove reaction by transaction ID, but didn't \
-                             find matching reaction in the reaction map"
-                    );
-                }
-            }
-            // Add the remote echo to the reaction_map
-            if let Some(event_id) = remote_echo_to_add {
-                let own_user_id = self.room_data_provider.own_user_id().to_owned();
-                state
-                    .reaction_map
-                    .insert((None, Some(event_id.to_owned())), (own_user_id, annotation.clone()));
-            };
+        if matches!(follow_up_action, ReactionAction::None) {
+            // We're done, update the timeline
+            state.reaction_state.remove(&annotation_key);
+            update_timeline_reaction(state, user_id, annotation, result)?;
```

Comment:
nit: Since there's only one branch in the `match` above that yields a `ReactionAction::None`, maybe this code could go up there, to avoid this `matches!` conditional?